packages/wrangler/src/miniflare-cli/assets.ts

```diff
@@ -28,465 +35,173 @@ export default async function generateASSETSBinding(options: Options) {
+					headers = parseHeaders(contents);
+					break;
+				}
+				case redirectsFile: {
+					log.log("_redirects modified. Re-evaluating...");
+					const contents = readFileSync(redirectsFile).toString();
+					redirects = parseRedirects(contents);
+					break;
 				}
 			}
 
-			if ((asset = getAsset(`/index.html`))) {
-				deconstructedResponse.body = serveAsset(asset);
-				deconstructedResponse.headers.set(
-					"Content-Type",
-					getType(asset) || "application/octet-stream"
-				);
-				return deconstructedResponse;
-			}
-
-			deconstructedResponse.status = 404;
-			return deconstructedResponse;
-		};
-
-		let asset;
-
-		if (assetName.endsWith("/")) {
-			if ((asset = getAsset(`${assetName}/index.html`))) {
-				deconstructedResponse.body = serveAsset(asset);
-				deconstructedResponse.headers.set(
-					"Content-Type",
-					getType(asset) || "application/octet-stream"
-				);
-				return deconstructedResponse;
-			} else if ((asset = getAsset(`${assetName.replace(/\/$/, ".html")}`))) {
-				deconstructedResponse.status = 301;
-				deconstructedResponse.headers.set(
-					"Location",
-					`${assetName.slice(0, -1)}${url.search}`
-				);
-				return deconstructedResponse;
-			}
+			metadata = createMetadataObject({
+				redirects,
+				headers,
+				logger: log.warn,
+			});
 		}
+	);
 
-		if (assetName.endsWith("/index")) {
-			deconstructedResponse.status = 301;
-			deconstructedResponse.headers.set(
-				"Location",
-				`${assetName.slice(0, -"index".length)}${url.search}`
-			);
-			return deconstructedResponse;
-		}
+	const generateResponse = async (request: Request) => {
+		const assetKeyEntryMap = new Map<string, string>();
+
+		return await generateHandler<string>({
+			request,
+			metadata: metadata as Metadata,
+			xServerEnvHeader: "dev",
+			logError: console.error,
+			findAssetEntryForPath: async (path) => {
+				const filepath = join(directory, path);
+
+				if (
+					existsSync(filepath) &&
+					lstatSync(filepath).isFile() &&
+					!ignoredFiles.includes(filepath)
+				) {
+					const hash = hashFile(filepath);
+					assetKeyEntryMap.set(hash, filepath);
+					return hash;
+				}
 
-		if ((asset = getAsset(assetName))) {
-			if (assetName.endsWith(".html")) {
-				const extensionlessPath = assetName.slice(0, -".html".length);
-				if (getAsset(extensionlessPath) || extensionlessPath === "/") {
-					deconstructedResponse.body = serveAsset(asset);
-					deconstructedResponse.headers.set(
-						"Content-Type",
-						getType(asset) || "application/octet-stream"
-					);
-					return deconstructedResponse;
+				return null;
+			},
+			getAssetKey: (assetEntry) => {
+				return assetEntry;
+			},
+			negotiateContent: (contentRequest) => {
+				let rawAcceptEncoding: string | undefined;
+				if (
+					contentRequest.cf &&
+					"clientAcceptEncoding" in contentRequest.cf &&
+					contentRequest.cf.clientAcceptEncoding
+				) {
+					rawAcceptEncoding = contentRequest.cf.clientAcceptEncoding as string;
```

Comment:
Am I right in thinking that this may resolve #370 if this code is also used in production?

Reply:
Not quite. We need to also assume a default "identity" accept-encoding if coming from `env.ASSETS.fetch()`, but that is something we'll be able to tackle shortly :)

I think it will probably help that comment on 165 though :)

Reply:
Progress!